### PR TITLE
feat: initial CLI commit

### PR DESCRIPTION
**Pull request by alexandreteles**
image

### CODE REVIEW

**Comment by alexandreteles** [comment]
Imagine being stopped by code coverage. It couldn't be me.

Anyway, to use the CLI, you will need the following:

### The obvious
Install all dependencies. Use `poetry install`, `poetry update`, or install from the requirements files.

### The not so obvious
Export a GitHub token to your env: `export GITHUB_TOKEN=my_amazing_token`

Then just ./cli.py

Everything else should be exploratory. The CLI uses the docstrings to generate help, so rewriting is trivial if something isn't clear. I will write instrumentation around it so it can be used from actions, but the design decision about how that can be achieved adequately from calling repositories needs more consideration.

I think the @ReVanced/infrastructure team also lacks people with experience in writing Github actions, so @PalmDevs, I might need your help with this.

File: cli.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    import uvloop
+
+    uvloop.install()
+except ImportError:
+    logger.info("Could not import uvloop. Using asyncio event loop.")
+
+
+class CLI(object):
+    """
+    CLI for the ReVanced API.
```
Comment (oSumAtrIX) [inline cli.py:24]:
```suggestion
    CLI for ReVanced API.
```

File: cli.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    def __gen_filtered_repositories(self) -> list[str]:
+        return list(
+            filter(
+                lambda repo: repo not in config.static_api_ignore_repos,
+                config.compat_repositories,
+            )
+        )
+
+    def get_filtered_repositories(self) -> list[str]:
```
Comment (oSumAtrIX) [inline cli.py:101]: Missing comment

File: cli.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                    attr = getattr(module, attr_name)
+                    if (
+                        callable(attr)
+                        and hasattr(attr, "is_cli_helper")
+                        and attr.is_cli_helper
+                    ):
+                        setattr(self, attr.cli_name or attr_name, attr)
+                        self.__imported_methods.append(attr.cli_name or attr_name)
+
+    def helper(self, name=None):
```
Comment (oSumAtrIX) [inline cli.py:47]: Missing comment

File: api/manager.py [hunk 4]
```diff
@@ -59,3 +64,58 @@ async def custom_sources(request: Request) -> JSONResponse:
     )
 
     return json(data, status=200)
+
+
+@cli.helper(name="manager")
+async def cli_root(bootstrap=None, custom_source=None, dev=False) -> dict[str, dict]:
+    """
+    Returns a dictionary containing a list of the main ReVanced tools.
```
Comment (oSumAtrIX) [inline api/manager.py:72]: This description is incorrect

File: api/compat.py [hunk 5]
```diff
@@ -76,12 +108,52 @@ async def contributors(request: Request) -> JSONResponse:
...
-            repositories=[
-                GithubRepository(owner=owner, name=repo) for repo in compat_repositories
-            ]
+    return json(await generate_contributors_data(), status=200)
+
+
+@cli.helper(name="tools")
+async def cli_tools() -> dict[str, dict]:
+    """
+    Retrieve a list of releases for a GitHub repository.
```
Comment (oSumAtrIX) [inline api/compat.py:117]: Incorrect description

File: api/compat.py [hunk 6]
```diff
@@ -76,12 +108,52 @@ async def contributors(request: Request) -> JSONResponse:
...
+            "Could not find path for tools. Did you generate the scaffolding?"
+        )
+
+    return ujson.dumps(data)
+
+
+@cli.helper(name="contributors")
+async def cli_contributors() -> dict[str, dict]:
+    """
+    Retrieve a list of releases for a GitHub repository.
```
Comment (oSumAtrIX) [inline api/compat.py:141]: Incorrect description

File: api/github.py [hunk 7]
```diff
@@ -61,6 +66,49 @@ async def list_releases(request: Request, repo: str) -> JSONResponse:
     return json(data, status=200)
 
 
+@cli.helper(name="releases")
+async def cli_releases(
+    repository: str, per_page: int = 100, page: int = 1
+) -> dict[str, dict]:
+    """
+    Retrieve a list of releases for a GitHub repository.
```
Comment (oSumAtrIX) [inline api/github.py:74]: Incorrect description

File: api/github.py [hunk 8]
```diff
@@ -161,6 +301,42 @@ async def get_contributors(request: Request, repo: str) -> JSONResponse:
     return json(data, status=200)
 
 
+@cli.helper(name="project-contributors")
```
Comment (oSumAtrIX) [inline api/github.py:304]: Simplify to "contributors"

File: api/github.py [hunk 9]
```diff
@@ -218,3 +440,31 @@ async def get_team_members(request: Request) -> JSONResponse:
     }
 
     return json(data, status=200)
+
+
+@cli.helper(name="team-members")
```
Comment (oSumAtrIX) [inline api/github.py:445]: Simplify to "team"

File: api/compat.py [hunk 10]
```diff
@@ -76,12 +108,52 @@ async def contributors(request: Request) -> JSONResponse:
...
 
-    data: dict[str, list] = {
-        "repositories": await github_backend.compat_get_contributors(
-            repositories=[
-                GithubRepository(owner=owner, name=repo) for repo in compat_repositories
-            ]
+    return json(await generate_contributors_data(), status=200)
+
+
+@cli.helper(name="tools")
```
Comment (oSumAtrIX) [inline api/compat.py:114]: May be out of scope for this PR but with the routes manager and patches this route is obsolete

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
- Are internal types such as JSONResponse or "dictionary" to be documented in places such as the command spec instead of saying "A json string" or "key value pair"?
- I have not reviewed all CLI command descriptions but many were incomplete/incorrect and need to be looked into
- Instead of having a release and releases command you should only provide "releases" with additional arguments to specify the tag (including the alias tag "latest") and a list of releases
- What is the purpose of the options "get_file_path", "get_filitered_repositories" or "scaffold"?
- May be out of scope but, the routes contributors and team could be merged with an argument to only retrieve the contributors that are present in the org (aka the team)

**Comment by oSumAtrIX** [comment]
Superseded by https://github.com/ReVanced/revanced-api/pull/169, thanks

**Pull request closed without merging**